apisix/plugins/aws-lambda.lua

```diff
@@ -0,0 +1,181 @@
+    headers["X-Amz-Date"] = amzdate
+
+    -- computing canonical uri
+    local canonical_uri = params.path:gsub("//", "/")
+    if canonical_uri ~= "/" then
+        if canonical_uri:sub(-1, -1) == "/" then
+            canonical_uri = canonical_uri:sub(1, -2)
+        end
+        if canonical_uri:sub(1, 1) ~= "/" then
+            canonical_uri = "/" .. canonical_uri
+        end
+    end
+
+    -- computing canonical query string
+    local canonical_qs = {}
+    for k, v in pairs(params.query) do
+        canonical_qs[#canonical_qs+1] = ngx.unescape_uri(k) .. "=" .. ngx.unescape_uri(v)
+    end
+
+    sort(canonical_qs)
+    canonical_qs = concat(canonical_qs, "&")
+
+    -- computing canonical and signed headers
+    local url_decoded = url.parse(conf.function_uri)
+    headers["host"] = url_decoded.host
+
+    local canonical_headers, signed_headers = {}, {}
+    for k, v in pairs(headers) do
+        k = k:lower()
+        if k ~= "connection" then
+            signed_headers[#signed_headers+1] = k
+            -- strip starting and trailing spaces including strip multiple spaces into single space
+            canonical_headers[k] =  v:gsub("^%s+", ""):gsub("%s+$", "")
+        end
+    end
+    sort(signed_headers)
+
+    for i = 1, #signed_headers do
+        local k = signed_headers[i]
+        canonical_headers[i] = k .. ":" .. canonical_headers[k] .. "\n"
+    end
+    canonical_headers = concat(canonical_headers, nil, 1, #signed_headers)
+    signed_headers = concat(signed_headers, ";")
+
+    -- combining elements to form the canonical request (step-1)
+    local canonical_request = params.method:upper() .. "\n"
+                        .. canonical_uri .. "\n"
+                        .. (canonical_qs or "") .. "\n"
+                        .. canonical_headers .. "\n"
+                        .. signed_headers .. "\n"
+                        .. sha256(params.body or "")
+
+    -- creating the string to sign for aws signature v4 (step-2)
+    local iam = conf.authorization.iam
+    local credential_scope = datestamp .. "/" .. iam.aws_region .. "/"
+                            .. iam.service .. "/aws4_request"
+    local string_to_sign = ALGO .. "\n"
+                        .. amzdate .. "\n"
+                        .. credential_scope .. "\n"
+                        .. sha256(canonical_request)
+
+    -- calculate the signature (step-3)
+    local signature_key = getSignatureKey(iam.secretkey, datestamp, iam.aws_region, iam.service)
+    local signature = hex.to_hex(hmac256(signature_key, string_to_sign))
+
+    -- add info to the headers (step-4)
+    headers["authorization"] = ALGO .. " Credential=" .. iam.accesskey
+                            .. "/" .. credential_scope
+                            .. ", SignedHeaders=" .. signed_headers
+                            .. ", Signature=" .. signature
+end
+
+
+return require("apisix.plugins.serverless.generic-upstream")(plugin_name,
```

Comment:
it is not easy to read.

another style:

```lua
local serverless_obj = require("apisix.plugins.serverless.generic-upstream")
local aws_obj = serverless_obj.new({
    name = plugin_name,
    version = plugin_version,
    priority = priority,
    ... ...
})
```

Reply:
Nice. Updated accordingly